cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.planner;
+
+import co.cask.cdap.etl.proto.Connection;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Sets;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * A special DAG that can insert connector nodes into a normal dag based on some rules.
+ * A connector node is a boundary at which the dag can be split into smaller dags.
+ */
+public class ConnectorDag extends Dag {
+  private final Set<String> reduceNodes;
+  private final Set<String> connectors;
+
+  public ConnectorDag(Collection<Connection> connections) {
+    this(connections, ImmutableSet.<String>of());
+  }
+
+  public ConnectorDag(Collection<Connection> connections, Set<String> reduceNodes) {
+    this(connections, reduceNodes, ImmutableSet.<String>of());
+  }
+
+  public ConnectorDag(Collection<Connection> connections, Set<String> reduceNodes, Set<String> connectors) {
+    super(connections);
+    this.reduceNodes = ImmutableSet.copyOf(reduceNodes);
+    this.connectors = new HashSet<>(connectors);
+  }
+
+  /**
+   * Insert connector nodes into the dag.
+   *
+   * A connector node is a boundary at which the pipeline can be split into sub dags.
+   * It is treated as a sink within one subdag and as a source in another subdag.
+   * A connector is inserted in front of a reduce node (aggregator plugin type, etc)
+   * when there is a path from some source to one or more reduce nodes or sinks.
+   * This is required because in a single mapper, we can't write to both a sink and do a reduce.
+   * We also can't reduce on multiple keys in the same mapper.
```

Comment:
this is confusing. 
- Can we reduce at all in a mapper? 
- are you saying that each sub-dag will still be implemented by a map-only job?